File: backend/pkg/sdjwt/disclosure.go
package sdjwt

import (
	"crypto/rand"
	"fmt"
	"io"
)

// The claim names RFC 9901 reserves for conveying digests (§4.1 rule 7).
const (
	// sdClaim holds the digests of disclosed object properties (§4.2.4.1).
	sdClaim = "_sd"
	// sdAlgClaim names the hash algorithm (§4.1.1).
	sdAlgClaim = "_sd_alg"
	// arrayDigestKey is the sole key of the object that stands in for a hidden
	// array element (§4.2.4.2). Three period characters, chosen because an
	// ellipsis reads as omitted content.
	arrayDigestKey = "..."
)

// saltEntropyBytes is the width of a generated salt. RFC 9901 §9.3 RECOMMENDS
// at least 128 bits of randomness, and notes that base64url-encoding exactly
// that much is the straightforward way to produce the string the format wants.
//
// The salt is the whole of the hiding property: it is what stops a Verifier
// holding an undisclosed digest from enumerating the plausible values of a
// claim — "true", "false", a country code — and hashing each until one
// matches.
const saltEntropyBytes = 16

// Disclosure is one selectively disclosable value: a salt, an optional claim
// name, and the value itself.
//
// The encoded string is the authoritative form, not a rendering of the fields.
// RFC 9901 computes the digest over the base64url string itself precisely so
// that no canonicalisation is needed — whitespace, Unicode escaping and
// property order are all free, and whichever variant the Issuer chose is what
// the signature commits to. Re-encoding the fields could produce a different
// string with a different digest, so a Disclosure carries what it was given
// and hands that back from String.
//
// The zero value is not usable; build one with NewObjectDisclosure,
// NewArrayDisclosure or ParseDisclosure.
type Disclosure struct {
	encoded string
	salt    string
	name    string
	// named distinguishes a three-element object-property Disclosure from a
	// two-element array-element one. A name of "" is a legal, if odd, claim
	// name, so the presence of the name cannot be inferred from its value.
	named bool
	value any
}

// NewObjectDisclosure creates a Disclosure for an object property (§4.2.1):
// the three-element array [salt, name, value].
//
// value may be any JSON-encodable Go value. It is encoded immediately and the
// Disclosure is then re-read from that encoding, so the value this Disclosure
// reports is the value a Verifier will see — json.Number for numbers,
// map[string]any for objects — rather than whatever Go type went in. One
// representation everywhere is worth the extra round trip; two would mean
// issuance and verification could disagree about what was disclosed.
func NewObjectDisclosure(salt, name string, value any) (Disclosure, error) {
	if name == sdClaim || name == arrayDigestKey {
		return Disclosure{}, fmt.Errorf("%w: claim name %q is reserved", ErrReservedClaim, name)
	}
	return newDisclosure([]any{salt, name, value})
}

// NewArrayDisclosure creates a Disclosure for an array element (§4.2.2): the
// two-element array [salt, value]. An array element has no name — its place in
// the array is its identity.
func NewArrayDisclosure(salt string, value any) (Disclosure, error) {
	return newDisclosure([]any{salt, value})
}

// newDisclosure encodes the array and re-parses it, so that every Disclosure
// in existence came out of ParseDisclosure and none can disagree with its own
// encoding.
func newDisclosure(elements []any) (Disclosure, error) {
	encoded, err := encodeJSON(elements)
	if err != nil {
		return Disclosure{}, fmt.Errorf("encode disclosure: %w", err)
	}
	return ParseDisclosure(b64.EncodeToString(encoded))
}

// ParseDisclosure reads a Disclosure from its base64url-encoded wire form.
func ParseDisclosure(encoded string) (Disclosure, error) {
	raw, err := b64.DecodeString(encoded)
	if err != nil {
		return Disclosure{}, fmt.Errorf("%w: not base64url: %w", ErrMalformedDisclosure, err)
	}
	decoded, err := decodeJSON(raw)
	if err != nil {
		return Disclosure{}, fmt.Errorf("%w: not JSON: %w", ErrMalformedDisclosure, err)
	}
	elements, ok := decoded.([]any)
	if !ok {
		return Disclosure{}, fmt.Errorf("%w: expected a JSON array, got %T", ErrMalformedDisclosure, decoded)
	}

	d := Disclosure{encoded: encoded}
	switch len(elements) {
	case 2:
		d.value = elements[1]
	case 3:
		name, ok := elements[1].(string)
		if !ok {
			return Disclosure{}, fmt.Errorf("%w: claim name must be a string, got %T",
				ErrMalformedDisclosure, elements[1])
		}
		d.named, d.name, d.value = true, name, elements[2]
	default:
		return Disclosure{}, fmt.Errorf("%w: expected 2 or 3 elements, got %d",
			ErrMalformedDisclosure, len(elements))
	}

	salt, ok := elements[0].(string)
	if !ok {
		return Disclosure{}, fmt.Errorf("%w: salt must be a string, got %T",
			ErrMalformedDisclosure, elements[0])
	}
	d.salt = salt
	return d, nil
}

// String returns the base64url wire form: exactly the bytes a digest is
// computed over, and exactly the bytes that appear between tildes.
func (d Disclosure) String() string { return d.encoded }

// Salt returns the salt.
func (d Disclosure) Salt() string { return d.salt }

// Name returns the claim name and true for an object-property Disclosure, and
// "", false for an array-element one.
func (d Disclosure) Name() (string, bool) { return d.name, d.named }

// Value returns the disclosed value, decoded as this package decodes all JSON:
// objects as map[string]any, arrays as []any, numbers as json.Number.
func (d Disclosure) Value() any { return d.value }

// Digest returns the base64url-encoded digest of this Disclosure under h —
// the string that appears in an _sd array or behind a "..." key.
func (d Disclosure) Digest(h HashAlg) (string, error) { return h.Digest(d.encoded) }

// newSalt reads saltEntropyBytes from r and base64url-encodes them.
//
// r is a parameter so that a test can pin the salts and get a byte-identical
// SD-JWT out of a byte-identical input; production passes nil and gets
// crypto/rand. math/rand is banned module-wide, and this is one of the places
// that ban is protecting: a predictable salt makes the digest of a low-entropy
// claim trivially reversible.
func newSalt(r io.Reader) (string, error) {
	if r == nil {
		r = rand.Reader
	}
	buf := make([]byte, saltEntropyBytes)
	if _, err := io.ReadFull(r, buf); err != nil {
		return "", fmt.Errorf("read salt: %w", err)
	}
	return b64.EncodeToString(buf), nil
}

File: frontend/src/sdjwt/disclosure.ts
/**
 * One disclosure: the salt, the value, and — for an object property — the name.
 *
 * # The encoded string is the disclosure
 *
 * `encoded` is not a rendering of the other fields, it is what the digest is
 * computed over, and the fields are what was read out of it. RFC 9901 §4.2.3
 * hashes the base64url string precisely so that no canonicalisation is needed:
 * whitespace, Unicode escaping and property order are all free, and whichever
 * variant the issuer chose is what the signature commits to. The RFC prints the
 * consequence as a worked example — the same `family_name` disclosure with the
 * umlaut written literally and written as a six-character JSON escape is one
 * value, two strings and two different digests.
 *
 * Re-encoding these fields and hashing *that* is the classic bug in this format
 * and it is worth naming, because of how it fails: not loudly, but with every
 * digest matching nothing, which the resolver reports as every claim having
 * been withheld. A screen then renders a mandate that looks minimised rather
 * than one that was misread.
 *
 * # Two shapes, and reading one as the other loses the value
 *
 * `[salt, name, value]` is an object property (§4.2.1); `[salt, value]` is an
 * array element (§4.2.2), which has no name because its place in the array is
 * its identity. They are a discriminated union here rather than an optional
 * `name`, because the failure they invite is silent: read a two-element
 * disclosure as a three-element one and the *value* lands in the name slot and
 * the value reads as `undefined`. Under `strict` a union makes that
 * unexpressible — nothing can reach `.name` without narrowing `kind` first.
 *
 * A name of `""` is a legal if odd claim name, so the presence of a name cannot
 * be inferred from its value either.
 */

import { decodeBase64UrlText } from "./base64url";
import { SdJwtError } from "./errors";

/** An object property's disclosure: `[salt, name, value]`. */
export interface ObjectDisclosure {
  readonly kind: "object";
  /** The base64url string, and the exact input to the digest. */
  readonly encoded: string;
  readonly salt: string;
  readonly name: string;
  readonly value: unknown;
}

/** An array element's disclosure: `[salt, value]`. */
export interface ArrayDisclosure {
  readonly kind: "array";
  /** The base64url string, and the exact input to the digest. */
  readonly encoded: string;
  readonly salt: string;
  readonly value: unknown;
}

export type Disclosure = ObjectDisclosure | ArrayDisclosure;

/**
 * Reads a disclosure from its base64url wire form.
 *
 * @param what names the position — which hop, which component — since a chain
 * carries several runs of these and "at position 2" counted from the wrong one
 * points at the wrong tilde.
 */
export function parseDisclosure(encoded: string, what: string): Disclosure {
  const text = decodeBase64UrlText(encoded, what);

  let decoded: unknown;
  try {
    decoded = JSON.parse(text);
  } catch (cause) {
    throw new SdJwtError("malformed_disclosure", `${what} is not JSON`, { cause });
  }
  if (!Array.isArray(decoded)) {
    throw new SdJwtError("malformed_disclosure", `${what} is not a JSON array`);
  }

  if (decoded.length !== 2 && decoded.length !== 3) {
    throw new SdJwtError(
      "malformed_disclosure",
      `${what} has ${decoded.length} elements; a disclosure has two — [salt, value] ` +
        `for an array element — or three — [salt, name, value] for an object property`,
    );
  }

  const salt = decoded[0];
  if (typeof salt !== "string") {
    throw new SdJwtError("malformed_disclosure", `${what}: the salt is not a string`);
  }

  if (decoded.length === 2) {
    return { kind: "array", encoded, salt, value: decoded[1] };
  }
  const name = decoded[1];
  if (typeof name !== "string") {
    throw new SdJwtError("malformed_disclosure", `${what}: the claim name is not a string`);
  }
  return { kind: "object", encoded, salt, name, value: decoded[2] };
}
